### PR TITLE
MST/transport: send source public key in MstState

### PR DESCRIPTION
**Pull request by MBoldyrev**
Also make state storage identify peers by their public keys.

### Description of the Change

Before this change MstState messages consisted of a transactions batch and the destination peer description. The receiving part interpreted the peer description as the source peer which the message came from, which is obviously an error.

In this patch, I replaced the peer description part of MstState message with the source peer's public key, which the receiving part uses to identify the receiver. Note that this change is not addressing the authentication of message.

### Benefits

Fixed the MST State propagation.

### Possible Drawbacks 

Yet to be found.

### CODE REVIEW

File: irohad/multi_sig_transactions/impl/mst_processor_impl.cpp [hunk 1]
```diff
@@ -13,14 +13,16 @@ namespace iroha {
       std::shared_ptr<iroha::network::MstTransport> transport,
       std::shared_ptr<MstStorage> storage,
       std::shared_ptr<PropagationStrategy> strategy,
-      std::shared_ptr<MstTimeProvider> time_provider)
+      std::shared_ptr<MstTimeProvider> time_provider,
+      shared_model::crypto::PublicKey my_key)
```
Comment (muratovv) [inline irohad/multi_sig_transactions/impl/mst_processor_impl.cpp]: Seems that own pub_key should be passed only in transport.

File: schema/mst.proto [hunk 2]
```diff
@@ -7,7 +7,7 @@ import "google/protobuf/empty.proto";
 
 message MstState {
     repeated iroha.protocol.Transaction transactions = 1;
-    iroha.protocol.Peer peer = 2;
+    bytes source_peer_key = 3;
```
Comment (muratovv) [inline schema/mst.proto]: Seems that the number should be `2`.

File: test/module/irohad/multi_sig_transactions/mst_processor_test.cpp [hunk 3]
```diff
@@ -48,6 +49,7 @@ class MstProcessorTest : public testing::Test {
   /// use effective implementation of storage
   std::shared_ptr<MstStorage> storage;
   std::shared_ptr<FairMstProcessor> mst_processor;
+  const shared_model::crypto::Keypair kMyKeypair = makeKey();
```
Comment (muratovv) [inline test/module/irohad/multi_sig_transactions/mst_processor_test.cpp]: No so reliable to create effective dependencies for components in unit-test better to rework it with mock.

Comment (MBoldyrev) [reply test/module/irohad/multi_sig_transactions/mst_processor_test.cpp]: I have done it the same way it is done in this file, using the same function and types. If it is wrong, then the entire thing should be changed in a separate patch.

File: irohad/network/mst_transport.hpp [hunk 4]
```diff
@@ -46,6 +46,7 @@ namespace iroha {
       /**
        * Share state with other peer
        * @param to - peer recipient of message
+       * @param src_key - the public key of source (this) peer
```
Comment (lebdron) [inline irohad/network/mst_transport.hpp]: Can be removed, since there is no such parameter

File: irohad/multi_sig_transactions/hash.hpp [hunk 5]
```diff
@@ -26,6 +26,14 @@ namespace iroha {
       std::size_t operator()(
           const std::shared_ptr<shared_model::interface::Peer> &obj) const;
```
Comment (lebdron) [inline irohad/multi_sig_transactions/hash.hpp]: PeerHasher can probably be removed now

File: irohad/multi_sig_transactions/storage/mst_storage.hpp [hunk 6]
```diff
@@ -43,7 +43,7 @@ namespace iroha {
      * General note: implementation of method covered by lock
      */
     StateUpdateResult apply(
-        const std::shared_ptr<shared_model::interface::Peer> &target_peer,
+        const shared_model::crypto::PublicKey &target_peer_key,
```
Comment (lebdron) [inline irohad/multi_sig_transactions/storage/mst_storage.hpp]: Please fix doc line 40

File: irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp [hunk 7]
```diff
@@ -87,19 +86,9 @@ grpc::Status MstTransportGrpc::SendState(
...
-          [&](expected::Value<std::unique_ptr<shared_model::interface::Peer>>
-                  &v) {
-            subscriber_.lock()->onNewState(std::move(v.value),
-                                           std::move(new_state));
-          },
-          [&](expected::Error<std::string> &e) {
-            async_call_->log_->info(e.error);
-          });
+  subscriber_.lock()->onNewState(
+      shared_model::crypto::PublicKey(request->source_peer_key()),
```
Comment (lebdron) [inline irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp]: `request->source_peer_key()` should be validated before creating a shared_model object.

File: irohad/multi_sig_transactions/storage/mst_storage.hpp (hunk 6 above)
Comment (MBoldyrev) [reply irohad/multi_sig_transactions/storage/mst_storage.hpp]: done

File: irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp [hunk 8]
```diff
@@ -86,8 +87,17 @@ grpc::Status MstTransportGrpc::SendState(
...
+  shared_model::crypto::PublicKey source_key(request->source_peer_key());
+  auto key_invalid_reason =
+      shared_model::validation::validatePubkey(source_key);
+  if (key_invalid_reason) {
+    async_call_->log_->info(
+        "Dropping received MST State due to invalid public key: {}",
+        *key_invalid_reason);
+  }
+
   subscriber_.lock()->onNewState(
```
Comment (muratovv) [inline irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp]: Seems that this line could be executed after the fail of `request::source_peer_key` validation.

**Comment by muratovv** [comment]
Also, pls add TODO(and task) for fixing dependencies of mst_processor_test

**Pull request merged**